### DISCUSSION TITLE
`offsetUnset` -> re-index?

### DISCUSSION THREAD

**Discussion by edalzell** [spatie/laravel-data#301] [Q&A]
Great package!

I'm storing a JSON array into a DB, and when I use `offsetUnset` it properly removes the item, but doesn't re-index so the stored JSON is wrong.

Any advice?

**Comment by edalzell** [comment]
Like I can solve it w/ getting the data as an array, splicing, then rehydrating, but seems a bit awkward.

This is what I'm doing now: 

```php
        $raw = $this->stations->toArray();

        array_splice($raw, $index, 1);

        $this->stations = new DataCollection(Station::class, $raw);
```

**Comment by rubenvanassche** [comment]
Hi @edalzell,

Do you have a complete code example where this is going wrong? Then I'll take a closer look, it's not completely clear for my what's going wrong.

**Comment by edalzell** [comment]
Sure thing Ruben, thanks for looking in to this.

I have a model, `Workout`, with casts like this:

```php
    protected $casts = [
        'scheduled_at' => 'datetime',
        'stations' => DataCollection::class.':'.Station::class,
        'type' => WorkoutType::class,
    ];
```

Station:
```php
class Station extends Data implements Wireable
{
    use WireableData;

    public function __construct(
        #[Required, IntegerType]
        public ?int $sets,

        #[Required, StringType]
        public ?string $title,

        #[Required, DataCollectionOf(Exercise::class)]
        public ?DataCollection $exercises
    ) {
    }
}
```

`toArray` on `stations` looks like this, right now:
```php
[
  [
    "sets" => 2,
    "title" => "Station 1",
    "exercises" => [
      [
        "id" => "bc6acc17-b792-4991-a54c-82fba0057aae",
        "activity" => "7x Quisquam deserunt sit ipsum.",
      ],
      [
        "id" => "ece742e4-1686-43bd-bc83-86aa0adb4f73",
        "activity" => "5x Aut est aperiam vel amet.",
      ],
    ],
  ],
  [
    "sets" => 3,
    "title" => "Station 2",
    "exercises" => [
      [
        "id" => "bc6acc17-b792-4991-a54c-82fba0057aae",
        "activity" => "7x Quisquam deserunt sit ipsum.",
      ],
      [
        "id" => "ece742e4-1686-43bd-bc83-86aa0adb4f73",
        "activity" => "5x Aut est aperiam vel amet.",
      ],
    ],
  ],
]
```

If I `offsetUnset` the first element then do `toArray` again, I get:

<img width="1758" alt="CleanShot 2022-12-23 at 09 39 20@2x" src="https://user-images.githubusercontent.com/6069653/209379564-8953ea29-5c4a-4a65-9bac-4e9fb7c31f9f.png">

See how there's an index now `1`? If that's there the JSON is stored with that as a key, instead of a normal array.

I understand why that's happening but maybe there should be a `removeItem` call or something that `values()` things to remove the indexes?

**Comment by rubenvanassche** [comment]
Hmmm I get what you're suggesting but this seems like correct behavior to me since it is how PHP works. As for the `removeItem` method. I don't think Laravel collections have such a feature and I want to follow those collections as close as possible.

You could always implement a custom DataCollection, and add such a method over there.

**Reply by edalzell** [reply]
Yup, you're right, I should be using something like `splice`. Perhaps I'll PR that.
